iceoryx_posh/include/iceoryx_posh/internal/popo/used_chunk_list.inl

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
+// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_POSH_POPO_USED_CHUNK_LIST_INL
+#define IOX_POSH_POPO_USED_CHUNK_LIST_INL
+
+namespace iox
+{
+namespace popo
+{
+template <uint32_t Capacity>
+constexpr typename UsedChunkList<Capacity>::DataElement_t UsedChunkList<Capacity>::DATA_ELEMENT_NULLPTR;
+
+template <uint32_t Capacity>
+UsedChunkList<Capacity>::UsedChunkList() noexcept
+{
+    static_assert(sizeof(DataElement_t) <= 8U, "The size of the data element type must not exceed 64 bit!");
+    static_assert(std::is_trivially_copyable<DataElement_t>::value,
+                  "The data element type must be trivially copyable!");
+
+    init();
+}
+
+template <uint32_t Capacity>
+bool UsedChunkList<Capacity>::insert(mepoo::SharedChunk chunk) noexcept
+{
+    if (freeSpaceInList())
+    {
+        // get next free entry after freelistHead
+        auto nextFree = m_listNodes[m_freeListHead];
+
+        // freeListHead is getting new usedListHead, next of this entry is updated to next in usedList
+        m_listNodes[m_freeListHead] = m_usedListHead;
+        m_usedListHead = m_freeListHead;
+
+        // store chunk mgmt ptr
+        rp::RelativePointer<mepoo::ChunkManagement> ptr = chunk.release();
+        auto segment = ptr.getId();
+        auto offset = ptr.getOffset();
+        cxx::Ensures(segment < DataElement_t::MAX_SEGMENT && "relative_ptr Id must fit into segment type!");
+        cxx::Ensures(offset < DataElement_t::MAX_OFFSET && "relative_ptr offset must fit into offset type!");
+        m_listData[m_usedListHead] = DataElement_t(static_cast<uint16_t>(segment), offset);
+
+        // set freeListHead to the next free entry
+        m_freeListHead = nextFree;
+
+        /// @todo can we do this cheaper with a global fence in cleanup?
```

Comment:
I we only use this from two contexts and can enforce that all writes of context 1 are made visible in context 2 (Roudi) when it needs to clean up then it should be sufficient to synchronize with a fence on cleanup. But this assumes that the values in the cache of the core where the application crashed on are still available. I do not know how this behaves across multiple applications, they might be cleaned up without being propagated to main memory or staying in cache.
The way it is should ensure at most one chunk (per list) can be leaked.